### PR TITLE
API-7729 advanced sort

### PR DESCRIPTION
**Pull request by jblefkove-va**
https://vajira.max.gov/browse/API-7729

Add support for advanced sorting of queries

See [DQ PR](https://github.com/department-of-veterans-affairs/health-apis-data-query/pull/743) for implementation of `Appointment?_sort=date`

### CODE REVIEW

File: src/main/java/gov/va/api/lighthouse/vulcan/SortRequest.java [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+                        .build();
+                  }
+                })
+            .collect(toList());
+    return builder().sorting(parameters).build();
+  }
+
+  @Value
+  @Builder
+  public static final class Parameter {
```
Comment (jhulbert-va) [inline src/main/java/gov/va/api/lighthouse/vulcan/SortRequest.java:45]: The naming on this feels a bit general. Maybe something like `SortableParameter`?

File: src/main/java/gov/va/api/lighthouse/vulcan/VulcanConfiguration.java [hunk 2]
```diff
@@ -47,16 +46,18 @@
...
 
-    @NotNull String countParameter;
+    @NonNull String countParameter;
 
     @Builder.Default int defaultCount = 10;
 
     @Builder.Default int maxCount = 20;
 
-    @NotNull Sort sort;
+    @NonNull Sort sort;
```
Comment (jhulbert-va) [inline src/main/java/gov/va/api/lighthouse/vulcan/VulcanConfiguration.java:55]: What about deprecating this and creating a `defaultSort` field? I think `sort` becomes kinda ambiguous with the addition of `sortableParameters`.

File: src/main/java/gov/va/api/lighthouse/vulcan/SortRequest.java (hunk 1 above)
Comment (jblefkove-va) [reply src/main/java/gov/va/api/lighthouse/vulcan/SortRequest.java:45]: It's an inner class, I think `SortRequest.Parameter` is okay

File: src/main/java/gov/va/api/lighthouse/vulcan/VulcanConfiguration.java (hunk 2 above)
Comment (jblefkove-va) [reply src/main/java/gov/va/api/lighthouse/vulcan/VulcanConfiguration.java:55]: Will do

**Review by jhulbert-va — DISMISSED** [review]
LGTM

File: src/main/java/gov/va/api/lighthouse/vulcan/VulcanConfiguration.java [hunk 3]
```diff
@@ -47,19 +46,34 @@
...
+
+    public static final class PagingConfigurationBuilder {
+      /**
+       * Set default sort.
+       *
+       * @deprecated use {@link #sortDefault}
+       */
+      @SuppressWarnings("InlineMeSuggester")
+      @Deprecated(since = "2.0.4", forRemoval = true)
+      public PagingConfigurationBuilder sort(Sort s) {
```
Comment (bschofield-va) [inline src/main/java/gov/va/api/lighthouse/vulcan/VulcanConfiguration.java:74]: ❤️

File: src/main/java/gov/va/api/lighthouse/vulcan/VulcanConfiguration.java [hunk 4]
```diff
@@ -47,19 +46,34 @@
...
 
     @Builder.Default int maxCount = 20;
 
-    @NotNull Sort sort;
-
     @NonNull Vulcan.BaseUrlStrategy baseUrlStrategy;
 
+    @NonNull Sort sortDefault;
+
+    Function<SortRequest, Sort> sortableParameters;
```
Comment (bschofield-va) [inline src/main/java/gov/va/api/lighthouse/vulcan/VulcanConfiguration.java:59]: A `noSortableParameters()` utility function would make this 🥇

**Review by bschofield-va — DISMISSED** [review]

**Review by jhulbert-va — APPROVED** [review]

**Pull request merged**